htdp-lib/lang/private/sl-stepper-button.rkt

```diff
@@ -0,0 +1,236 @@
+         (prefix-in pc: mzlib/pconvert)
+         (prefix-in ic: mrlib/image-core)
+         mrlib/cache-image-snip
+         lang/private/set-result
+         lang/stepper-language-interface
+         (only-in deinprogramm/signature/signature signature? signature-name)
+         stepper/drracket-button)
+
+(define (sl-stepper-drracket-button options)
+  (stepper-drracket-button (new sl-stepper-language%) (options->settings options)))
+
+(define-logger stepper)
+
+(define-struct simple-settings (show-sharing
+                                insert-newlines
+                                true/false/empty-as-ids?
+                                abbreviate-cons-as-list
+                                use-function-output-syntax?))
+
+(define (present? key options)
+  (and (memq key options) #t))
+
+(define (options->settings options)
+  (make-simple-settings (present? 'show-sharing options)
+                        #t #f ; vestige from old levels
+                        (present? 'abbreviate-cons-as-list options)
+                        (present? 'use-function-output-syntax options)))
+   
+(define sl-stepper-language%
+  (class* object% (stepper-language<%>)
+    (public stepper:supported?)
+    (define (stepper:supported?) #t)
+
+    (public stepper:enable-let-lifting?)
+    (define (stepper:enable-let-lifting?) #f)
+
+    ;; these next three parameters should be overridden by
+    ;; the language definition to match the way that the language
+    ;; wants these values printed.
+    (public stepper:show-lambdas-as-lambdas?)
+    (define (stepper:show-lambdas-as-lambdas?) #t)
+
+    (public stepper:show-inexactness?)
+    (define (stepper:show-inexactness?) #t)
+
+    (public stepper:print-boolean-long-form?)
+    (define (stepper:print-boolean-long-form?) #t)
+
+    (public stepper:show-consumed-and/or-clauses?)
+    (define (stepper:show-consumed-and/or-clauses?) #t)
+
+    (public stepper:render-to-sexp)
+    (define (stepper:render-to-sexp val settings language-level)
+      (when (boolean? val)
+        (log-stepper-debug "render-to-sexp got a boolean: ~v\n" val))
+      (or (and (procedure? val)
+               (object-name val))
+          (parameterize ([pretty-print-show-inexactness (stepper:show-inexactness?)]
+                         [current-print-convert-hook stepper-print-convert-hook])
+            (call-with-values
+             (lambda ()
+               ;; I'm not sure that these print settings actually need to be set...
+               ;; or... that they need to be set *here*. They might need to be set
+               ;; when the pretty-print to the actual width occurs. That is, when
+               ;; they get converted to strings...
+               ;; try removing this wrapper when things are working. (2015-10-22)
+               (call-with-print-settings
+                language-level
+                settings
+                (lambda ()
+                  (simple-module-based-language-convert-value
```

Comment:
No, you're right. I did this primarily to reduce risk, i.e. to avoid changes elsewhere.

The idea is to then finally consolidate all that duplicated rendering code for the next cycle. But doing it now would mean changes to other parts of the code, which I'd like t avoid.

Reply:
So the idea is that we increase ugliness now, and go back and fix it after the release? I'm worried that I'm going to drop the ball on this. I see that it's somewhat unfair for me to hold up your change because of my low opinion of myself...

Can we set some kind of timeline for this cleanup? Do you have time to fix this (on the master branch, not on the release) during January?

Reply:
Great! January is probably a bit tight, but I can commit to mid-February.

Many thanks for your work on this!

Reply:
Let's try to do these fixes at the start of a release cycle, rather than the end of one?

Reply:
> Let's try to do these fixes at the start of a release cycle, rather than the end of one?

That's what @jbclements and I are saying. (I think. Right?)

Reply:
Mike, do you have time to talk? I'm still confused, and my eyes are starting to hurt. I'm hoping you can explain your changes to me and help me understand which pieces go where. I've been staring at this for hours and I don't have a clear picture of what does what. I also feel like I have no choice but to move into the testing phase early tomorrow, which means that getting this pull request in will require approval from a bunch more people. I know this is 100% my fault for not reading this code earlier, and I'm feeling quite guilty. I'm also really not very happy about putting code in an "8.0" release that we know is ugly and intermediate. I'm scared of getting burned.